src/ASM/ASMs2Dmx.C

```diff
@@ -277,60 +278,59 @@ bool ASMs2Dmx::generateFEMTopology ()
-      for (i1 = 0; i1 < it->numCoefs_u(); i1++)
+    for (int i2 = 0; i2 < it->numCoefs_v(); i2++)
+      for (int i1 = 0; i1 < it->numCoefs_u(); i1++)
       {
         myNodeInd[inod].I = i1;
         myNodeInd[inod].J = i2;
         myMLGN[inod++]    = ++gNod;
       }
 
-  int firstItgNode = this->getFirstItgElmNode();
-  int nElmNode = std::accumulate(elem_size.begin(), elem_size.end(), 0);
+  const size_t firstItgElmNode = this->getFirstItgElmNode();
+  const size_t nElmNode = std::accumulate(elem_size.begin(), elem_size.end(), 0);
 
   // Create nodal connectivities for bases
+  auto&& enumerate_elem = [this,&iel](const Go::SplineSurface& spline,
+                                      size_t elm_node, int last_node)
+  {
+    for (int j = spline.order_v() - 1; j >= 0; j--)
+      for (int i = spline.order_u() - 1; i >= 0; i--)
+        myMNPC[iel][elm_node++] = last_node - spline.numCoefs_u()*j - i;
+  };
+
   inod = std::accumulate(nb.begin(),nb.begin()+geoBasis-1,0u);
-  Go::SplineSurface* b = m_basis[geoBasis-1].get();
-  auto knotv = b->basis_v().begin();
-  for (i2 = 1; i2 <= b->numCoefs_v(); i2++, ++knotv) {
-    auto knotu = b->basis_u().begin();
-    for (i1 = 1; i1 <= b->numCoefs_u(); i1++, inod++, ++knotu)
-      if (i1 >= b->order_u() && i2 >= b->order_v())
-      {
-        if (b->knotSpan(0,i1-1) > 0.0)
-          if (b->knotSpan(1,i2-1) > 0.0)
-          {
-            myMLGE[iel] = ++gEl; // global element number over all patches
-
-            int lnod = firstItgNode;
-            myMNPC[iel].resize(nElmNode,0);
-
-            for (j2 = b->order_v()-1; j2 >= 0; j2--)
-              for (j1 = b->order_u()-1; j1 >= 0; j1--)
-                myMNPC[iel][lnod++] = inod - b->numCoefs_u()*j2 - j1;
-
-            // find knotspan spanning element for other bases
-            lnod = 0;
-            size_t basis_ofs = 0;
-            for (size_t bas = 0; bas < m_basis.size(); ++bas) {
-              if (bas != static_cast<size_t>(geoBasis-1)) {
-                double ku = *knotu;
-                double kv = *knotv;
-                int bknotu = m_basis[bas]->basis_u().knotIntervalFuzzy(ku);
-                int bknotv = m_basis[bas]->basis_v().knotIntervalFuzzy(kv);
-                size_t iinod = basis_ofs + bknotv*m_basis[bas]->numCoefs_u() + bknotu;
-                for (j2 = m_basis[bas]->order_v()-1; j2 >= 0; j2--)
-                  for (j1 = m_basis[bas]->order_u()-1; j1 >= 0; j1--)
-                    myMNPC[iel][lnod++] = iinod - m_basis[bas]->numCoefs_u()*j2 - j1;
-              } else
-                lnod += elem_size[bas];
-              basis_ofs += nb[bas];
+  auto knotv = itg->basis_v().begin();
+  for (int i2 = 1; i2 <= itg->numCoefs_v(); i2++, ++knotv) {
+    auto knotu = itg->basis_u().begin();
+    for (int i1 = 1; i1 <= itg->numCoefs_u(); i1++, inod++, ++knotu)
+      if (i1 >= itg->order_u() && i2 >= itg->order_v()) {
+        if (itg->knotSpan(0,i1-1) > 0.0 && itg->knotSpan(1,i2-1) > 0.0) {
```

Comment:
Can maybe combine the  two if statements into a single one. Or even better, move the check on `i2` and `knotSpan(1,i2-1)` outside the inner-most loop, since it will evaluate to the the same each time, thereby saving some looping (microoptimization, think the results will be the same but needs to be verified), but you need to add a `else inod += itg->numCoefs_u();` though to compensate for the `inod++`

Reply:
while i can do the first part, it's .. tricky to the combine ifs. reason is we need to increase iel even if the knotspan is empty.